### PR TITLE
Fix EZP-22408: Deleting a content does not remove references to this content in relation(list) fields

### PR DESCRIPTION
**Pull request by pspanja**
This PR resolves issue https://jira.ez.no/browse/EZP-22408

This implements cleanup script.
For more details see `ezpublish-kernel` PR for the same issue at https://github.com/ezsystems/ezpublish-kernel/pull/993.

### CODE REVIEW

**Comment by andrerom** [comment]
@pspanja this seems to have been forgotten, and issue is closed as fixed in the meantime.
This can still be made part of a 5.4 sub release with update instructions.

**Comment by pspanja** [comment]
Rebased, still valid for review :)
Ping @andrerom @bdunogier @glye @lolautruche @yannickroger

**Comment by lolautruche** [comment]
+1

File: update/common/scripts/5.4/cleanupfieldvaluerelations.php [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        $result = $db->query(
+            "UPDATE ezcontentobject_attribute
+             LEFT JOIN ezcontentobject ON ezcontentobject_attribute.data_int = ezcontentobject.id
+             SET data_int = NULL, sort_key_int = 0
+             WHERE ezcontentobject_attribute.data_type_string = 'ezobjectrelation' AND
+                   ezcontentobject_attribute.data_int IS NOT NULL AND
+                   ezcontentobject.id IS NULL"
+        );
+    }
+    else if ( $db->databaseName() == 'postgresql' )
```
Comment (bdunogier) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: Anything for Oracle ? :-)

Comment (pspanja) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: Nope, AFAIK we don't support it :)

File: update/common/scripts/5.4/cleanupfieldvaluerelations.php [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        sleep( $optIterationSleep );
+        $pass++;
+        $relationListArray = loadRelationListRows( $pass );
+    }
+
+    $cli->output( "Updated total of {$relationListUpdatedCount} RelationList field values." );
+}
+
+$cli->output();
+$cli->output( "Processing RelationList field values..." );
```
Comment (bdunogier) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: You could actually skip this one entirely if there are no rows to process :-)

File: update/common/scripts/5.4/cleanupfieldvaluerelations.php [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+if ( $relationListCount > 0 )
+{
+    processRelationListRows();
+}
+else
+{
+    $cli->output( "There are not RelationList field values to process." );
```
Comment (bdunogier) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: s/not/no/

File: update/common/scripts/5.4/cleanupfieldvaluerelations.php (hunk 2 above)
Comment (pspanja) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: Moved to the `if` block below and rebased.

File: update/common/scripts/5.4/cleanupfieldvaluerelations.php (hunk 3 above)
Comment (pspanja) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: Fixed and rebased.

File: update/common/scripts/5.4/cleanupfieldvaluerelations.php [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+$cli->output();
+$cli->output( "Processing Relation field values..." );
+
+if ( $relationCount > 0 )
+{
+    processRelationRows();
+}
+else
+{
+    $cli->output( "There are not Relation field values to process." );
```
Comment (glye) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: One more :)

**Comment by glye** [comment]
+1 afaict other than speling comment.

File: update/common/scripts/5.4/cleanupfieldvaluerelations.php (hunk 4 above)
Comment (pspanja) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: Fixed :)

**Comment by lolautruche** [comment]
AFAICS, you've got enough +1 :wink:

**Comment by pspanja** [comment]
Thanks for the reminder :wink:

File: update/common/scripts/5.4/cleanupfieldvaluerelations.php [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+{
+    $cli->output();
+    $cli->warning( "You have 30 seconds to break the script before actual processing starts (press Ctrl-C)." );
+    $cli->warning( "Execute the script with '-n' switch to skip this delay." );
+    sleep( 30 );
+}
+
+function processRelationRows()
+{
+    global $db, $cli, $relationCount;
```
Comment (emodric) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: Isn't this gonna cause issues when running through ezpublish:legacy:script?

Ref: https://jira.ez.no/browse/EZP-21913

Comment (pspanja) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: Yes indeed it will. Thanks @emodric, I'll fix it.

Comment (pspanja) [inline update/common/scripts/5.4/cleanupfieldvaluerelations.php]: Fixed directly in https://github.com/ezsystems/ezpublish-legacy/commit/f5e28c0394203b410fdba583bb752717fc5da891.

**Pull request merged**